src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileMetadataService.scala

```diff
@@ -70,7 +68,8 @@ class FileMetadataService(
       fileIds: Set[UUID] = existingFileRows.toFileTypeIds
       _ <- fileMetadataRepository.deleteFileMetadata(fileIds, distinctPropertyNames)
       addedRows <- fileMetadataRepository.addFileMetadata(generateFileMetadataRows(fileIds, distinctMetadataProperties, userId))
-      _ <- validateFileMetadataService.validateAdditionalMetadata(uniqueFileIds.toSet, input.consignmentId, distinctPropertyNames)
+      fileStatuses <- validateFileMetadataService.validateAdditionalMetadata(uniqueFileIds.toSet, input.consignmentId, distinctPropertyNames)
+      _ <- consignmentStatusService.updateMetadataConsignmentStatus(consignmentId, fileStatuses, List(DescriptiveMetadata, ClosureMetadata))
```

Comment:
I'm not sure if you already considered the following scenarios but it's worth adding unit tests:

Update:

1. Set `Incomplete` consignment status if one of the existing file's status is `Incomplete`
2. Set `Complete` consignment status if none of the existing files' statuses are `Incomplete`

Delete:
1. Set `Incomplete` consignment status if one of the existing file's status is `Incomplete`
2. Set `Complete` consignment status if one of the existing file's status is `Complete`
2. Set `NotEntered` consignment status if all of the existing files' statuses are `NotEntered`

One more thing, since we are not going to add a database trigger at the moment, I was thinking to do the update consignment job asynchronously because the update file metadata call is a bit slow as we are validating on the frontend and API too, and if we do the consignment status update task (which is not related to the update/delete action) then it will be very heavy considering if we have thousands of files for a consignment.

Reply:
Those scenarios are [covered in the unit tests](https://github.com/nationalarchives/tdr-consignment-api/pull/575/files#diff-8af430ff536b04425b75b9adbd9c69938ea74433c1230ab729656c72206e0f29R393) for the `updateMetadataConsignmentStatus` method  
I don't think there's much point in checking them again here. The unit tests for this method in `FileMetadataService` check that the `updateMetadataConsignmentStatus` which is enough I think. 

For the second point, it depends what you mean by asynchronously. If you mean taking the `Future` out of the for comprehension then that's not a good idea as it can lead to some odd behaviour down the line. There are things we can do with Akka actors and bits but we're trying to get rid of Akka so I don't want to get more embedded in their ecosystem. I'm not sure how else we'd do it.
The other problem with doing this asynchronously is that there is no way for the frontend to know if there's a failure so if the insert fails, the data will be inconsistent and the user will see the wrong status.

Reply:
I've added in tests for the add and delete metadata which check which status it returns if there is an existing file outside of the update with a particular status.
I've also changed the code so the tests work.